server/session/handler_item_stack_request.go

```diff
@@ -495,3 +571,194 @@ func (h *ItemStackRequestHandler) reject(id int32, s *Session) {
+		if satisfiedInputs == len(inputs) {
+			break
+		}
+
+		slot := i + offset
+		oldSt, err := s.ui.Item(int(slot))
+		if err != nil {
+			return false
+		}
+		if oldSt.Empty() {
+			// We should still up the satisfied inputs count if both stacks are empty.
+			if inputs[satisfiedInputs].Empty() {
+				satisfiedInputs++
+			}
+
+			continue
+		}
+
+		currentInputToMatch := inputs[satisfiedInputs]
+
+		// Items that apply to all types, so we just compare with the name and count.
+		if currentInputToMatch.AllTypes {
+			name, _ := oldSt.Item().EncodeItem()
+			otherName, _ := currentInputToMatch.Item().EncodeItem()
+			if name == otherName && oldSt.Count() >= currentInputToMatch.Count() {
+				satisfiedInputs++
+			}
+		} else {
+			if oldSt.Comparable(currentInputToMatch.Stack) {
+				satisfiedInputs++
+			}
+		}
+	}
+
+	return satisfiedInputs == len(inputs)
+}
+
+// removeInventoryInputs removes the inputs in the player inventory.
+func (h *ItemStackRequestHandler) removeInventoryInputs(inputs []recipe.Item, s *Session) error {
+	inputMap := h.inputMapFromInputs(inputs)
+
+	updateStack := func(container byte, slot byte, oldSt item.Stack) {
+		if oldSt.Empty() {
+			return
+		}
+
+		name, meta := oldSt.Item().EncodeItem()
+		if data, ok := inputMap[name]; ok {
+			if data.metadataValue == meta || data.allTypes {
+				if data.count > 0 {
+					targetRemoval := oldSt.Count()
+					if data.count < oldSt.Count() {
+						targetRemoval = data.count
+					}
+
+					st := oldSt.Grow(-targetRemoval)
+					h.setItemInSlot(protocol.StackRequestSlotInfo{
+						ContainerID:    container,
+						Slot:           slot,
+						StackNetworkID: item_id(st),
+					}, st, s)
+
+					data.count -= targetRemoval
+					inputMap[name] = data
+				}
+			}
+		}
+	}
+
+	for slot, oldSt := range s.inv.Items() {
+		updateStack(containerFullInventory, byte(slot), oldSt)
+	}
+
+	offset := s.craftingOffset()
+	for i := byte(0); i < s.craftingSize(); i++ {
+		slot := i + offset
+
+		oldSt, err := s.ui.Item(int(slot))
+		if err != nil {
+			return err
+		}
+
+		updateStack(containerCraftingGrid, slot, oldSt)
+	}
+
+	return nil
+}
+
+// removeGridInputs removes the inputs passed in the crafting grid.
+func (h *ItemStackRequestHandler) removeGridInputs(inputs []recipe.Item, s *Session) error {
+	offset := s.craftingOffset()
+
+	var index int
+	for i := byte(0); i < s.craftingSize(); i++ {
+		if index == len(inputs) {
+			break
+		}
+
+		slot := i + offset
+		oldSt, err := s.ui.Item(int(slot))
+		if err != nil {
+			return fmt.Errorf("expected item doesn't exist: " + err.Error())
+		}
+		if oldSt.Empty() {
+			// We should still up the index if the expected input is empty.
+			if inputs[index].Empty() {
+				index++
+			}
+
+			continue
+		}
+
+		st := oldSt.Grow(-inputs[index].Count())
+		h.setItemInSlot(protocol.StackRequestSlotInfo{
+			ContainerID:    containerCraftingGrid,
+			Slot:           slot,
+			StackNetworkID: item_id(st),
+		}, st, s)
+		index++
+	}
```

Comment:
All of this is a ton of extra code. Do you reckon you could compact it somehow?

Reply:
I'm not really sure how I could. Are there any portions that are necessarily standing out to you when you say that?